### PR TITLE
Search clean-up: separate permissions and logical filters

### PR DESCRIPTION
**Pull request by crisptrutski**
A few small extras lumped in.

- separated permissions and filter related collections where clauses.
- made the personal collection query much clearer.
- forked the interface used for adding collections clauses, to make new code less awkward.
- search.util is only used by legacy search, so moved it into the sub-namespace.
- enabled caching for the weights (eek, lots of queries otherwise) - I see the app is smart about cache busting.

### CODE REVIEW

File: src/metabase/public_settings.clj [hunk 1]
```diff
@@ -1052,7 +1052,6 @@ See [fonts](../configuring-metabase/fonts.md).")
 (defsetting experimental-search-weight-overrides
   (deferred-tru "Used to override weights used for search ranking")
   :visibility :internal
-  :cache?     false
```
Comment (crisptrutski) [inline src/metabase/public_settings.clj:1055]: This was a bad idea, it caused about 30 extra queries for a legacy search.

TIL that our distributed cache busting is quite smart.

File: src/metabase/models/table.clj [hunk 2]
```diff
@@ -311,7 +311,7 @@
                   :view-count    true
                   :created-at    true
                   :updated-at    true}
-   :search-terms [:name :description :display_name]
+   :search-terms [:name :display_name :description]
```
Comment (crisptrutski) [inline src/metabase/models/table.clj:314]: Reordered so primary terms come first. Foreshadowing the primary / secondary spec split.

File: src/metabase/search/filter.clj [hunk 3]
```diff
@@ -77,6 +78,31 @@
...
+(defn personal-collections-where-clause
+  "Build a clause limiting the entries to those (not) within or within personal collections, if relevant.
+  WARNING: this method queries the appdb, and its approach will get very slow when there are many users!"
+  [{filter-type :filter-items-in-personal-collection} collection-id-col]
+  (when filter-type
+    (let [parent-ids     (t2/select-pks-vec :model/Collection :personal_owner_id [:not= nil])
+          child-patterns (for [id parent-ids] (format "/%d/%%" id))]
+      (case filter-type
+        "only"
+        `[:or
```
Comment (crisptrutski) [inline src/metabase/search/filter.clj:90]: Haven't really used quasi quotes like this before, but seems like a nice way to splice list expressions deeply in nested collections.

File: src/metabase/search/postgres/core.clj [hunk 4]
```diff
@@ -81,14 +81,26 @@
...
 (defn- fulltext
   "Search purely using the index."
   [search-term & {:as search-ctx}]
   (when-not @#'search.index/initialized?
     (throw (ex-info "Search index is not initialized. Use [[init!]] to ensure it exists."
                     {:search-engine :postgres})))
-  (->> (let [base-query (search.index/search-query search-term search-ctx [:legacy_input])]
-         (search.permissions/add-collection-join-and-where-clauses base-query "search-index" search-ctx))
+  (->> (search.index/search-query search-term search-ctx [:legacy_input])
+       (add-collection-join-and-where-clauses search-ctx)
```
Comment (crisptrutski) [inline src/metabase/search/postgres/core.clj:103]: This felt good!

File: test/metabase/api/search_test.clj [hunk 5]
```diff
@@ -313,7 +313,7 @@
   (when (search/supports-index?)
     (testing "It can use an alternate search engine"
       (with-search-items-in-root-collection "test"
-        (let [resp (search-request :crowberto :q "test" :search_engine "fulltext")]
+        (let [resp (search-request :crowberto :q "test" :search_engine "fulltext" :limit 1)]
```
Comment (crisptrutski) [inline test/metabase/api/search_test.clj:316]: When testing with a large dev appdb `search-request` tries to parse the result before the API has finished streaming. We don't care about the results, so we don't need any more results.

File: src/metabase/search/filter.clj (hunk 3 above)
Comment (piranha) [reply src/metabase/search/filter.clj:90]: 😮

File: src/metabase/search/filter.clj [hunk 6]
```diff
@@ -77,6 +78,31 @@
 
 (defmethod where-clause* ::list [_ k v] [:in k v])
 
+(defn personal-collections-where-clause
```
Comment (piranha) [inline src/metabase/search/filter.clj:81]: I think this guy should go through the indexing phase w/o cutting corners like that

Comment (crisptrutski) [reply src/metabase/search/filter.clj:81]: Yeah agreed - tag every index in a personal collection with the user id.

That's a larger change I'd prefer to make once there's better testing - this code is still share-able and being covered by the legacy tests.

File: src/metabase/search/filter.clj (hunk 3 above)
Comment (piranha) [reply src/metabase/search/filter.clj:90]: It's surprising, but also readable, so I'm kind of ambivalent about it 😁

**Review by sansolovyov — APPROVED** [review]

**Pull request merged**